File: frontend/lib/features/transactions/domain/transaction.dart
enum PaymentMode { upi, card, cash, netbanking }

PaymentMode _modeFromString(String s) {
  return PaymentMode.values.firstWhere(
    (m) => m.name == s,
    orElse: () => PaymentMode.upi,
  );
}

/// A single transaction line.
/// [amountPaise] is negative for a spend and positive for a refund/credit,
/// matching the "refund reduces category total" edge case in the spec.
class Txn {
  final String id;
  final String merchantRaw;
  final String merchantName;
  final String categoryId;
  final int amountPaise;
  final DateTime at;
  final PaymentMode mode;

  const Txn({
    required this.id,
    required this.merchantRaw,
    required this.merchantName,
    required this.categoryId,
    required this.amountPaise,
    required this.at,
    required this.mode,
  });

  bool get isRefund => amountPaise > 0;

  Txn copyWith({String? categoryId}) => Txn(
        id: id,
        merchantRaw: merchantRaw,
        merchantName: merchantName,
        categoryId: categoryId ?? this.categoryId,
        amountPaise: amountPaise,
        at: at,
        mode: mode,
      );

  /// Normalises a raw merchant string ("SWIGGY*1234", "UBER   TRIP#99")
  /// so merchant rules match reliably even when the suffix changes.
  static String normalise(String raw) {
    final noSuffix = raw.split(RegExp(r'[*#]')).first;
    return noSuffix.trim().toUpperCase().replaceAll(RegExp(r'\s+'), ' ');
  }

  factory Txn.fromJson(Map<String, dynamic> json) => Txn(
        id: json['id'] as String,
        merchantRaw: json['merchantRaw'] as String,
        merchantName: json['merchantName'] as String,
        categoryId: json['category'] as String,
        amountPaise: json['amountPaise'] as int,
        at: DateTime.parse(json['at'] as String),
        mode: _modeFromString(json['mode'] as String),
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'merchantRaw': merchantRaw,
        'merchantName': merchantName,
        'category': categoryId,
        'amountPaise': amountPaise,
        'at': at.toUtc().toIso8601String(),
        'mode': mode.name,
      };
}

File: backend/src/models/Transaction.js
const mongoose = require('mongoose');

const PAYMENT_MODES = ['upi', 'card', 'cash', 'netbanking'];

// Mirrors features/transactions/domain/transaction.dart's `Txn` exactly:
// id, merchantRaw, merchantName, category, amountPaise, at, mode.
// amountPaise is negative for a spend, positive for a refund/credit —
// the sign convention the whole app (and the aggregation below) relies on.
const transactionSchema = new mongoose.Schema(
  {
    userId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true, index: true },
    merchantRaw: { type: String, required: true, trim: true },
    merchantName: { type: String, required: true, trim: true },
    // Normalised merchant key ("SWIGGY*1234" -> "SWIGGY"), used to match
    // MerchantRule entries and to group merchant insights. Computed on
    // save, never sent by the client.
    merchantKey: { type: String, required: true, index: true },
    category: { type: String, required: true, index: true }, // Category.id
    amountPaise: { type: Number, required: true },
    at: { type: Date, required: true },
    mode: { type: String, enum: PAYMENT_MODES, required: true },
    // Denormalised "YYYY-MM" of `at` (server-local/UTC), so every
    // month-scoped query (feed, summary, merchants) is a plain indexed
    // equality match instead of a date-range scan.
    month: { type: String, required: true, index: true },
  },
  { timestamps: true, versionKey: false }
);

transactionSchema.index({ userId: 1, month: 1, at: -1, _id: -1 });
transactionSchema.index({ userId: 1, merchantKey: 1 });

// Matches Txn.normalise() in the Flutter app bit-for-bit:
//   raw.split(/[*#]/).first.trim().toUpperCase().replaceAll(/\s+/, ' ')
function normaliseMerchant(raw) {
  const withoutSuffix = String(raw).split(/[*#]/)[0];
  return withoutSuffix.trim().toUpperCase().replace(/\s+/g, ' ');
}

function monthKeyOf(date) {
  const y = date.getUTCFullYear();
  const m = String(date.getUTCMonth() + 1).padStart(2, '0');
  return `${y}-${m}`;
}

transactionSchema.pre('validate', function preValidate(next) {
  if (this.merchantRaw) {
    this.merchantKey = normaliseMerchant(this.merchantRaw);
  }
  if (this.at) {
    this.month = monthKeyOf(new Date(this.at));
  }
  next();
});

// Shape that matches Txn.fromJson exactly: id, merchantRaw, merchantName,
// category, amountPaise, at (ISO-8601 UTC), mode.
transactionSchema.methods.toPublicJSON = function toPublicJSON() {
  return {
    id: this._id.toString(),
    merchantRaw: this.merchantRaw,
    merchantName: this.merchantName,
    category: this.category,
    amountPaise: this.amountPaise,
    at: this.at.toISOString(),
    mode: this.mode,
  };
};

transactionSchema.statics.normaliseMerchant = normaliseMerchant;
transactionSchema.statics.monthKeyOf = monthKeyOf;
transactionSchema.statics.PAYMENT_MODES = PAYMENT_MODES;

module.exports = mongoose.model('Transaction', transactionSchema);
